### PR TITLE
Add native incr support for GCSBackend

### PR DESCRIPTION
**Pull request by haimjether**
* Implement chord ref count on top of Google Firestore
 * Improve runtime and reduce amount of data read from GCS.
 * Add tests
 * Update docs

### CODE REVIEW

**Review by auvipy — CHANGES_REQUESTED** [review]
thanks for the patch! this will be really great addition. will you please also look at pypy failure and increase test coverage?

**Comment by haimjether** [comment]
> thanks for the patch! this will be really great addition. will you please also look at pypy failure and increase test coverage?

np, on it :)

**Comment by haimjether** [comment]
I struggle to fix the grpcio installation (required by `google-cloud-firestore`)  for pypy in the test image. 
A bit of comfort is that  it looks like a familiar issue:  https://github.com/grpc/grpc/issues/29517

FWIW, was able to build the `grpcio-1.66.1-pp310-pypy310_pp73-linux_x86_64.whl`  from sources in clean environment which I created (without tox being involved) on Ubuntu 24.04.  The whole thing takes ~8.5 minutes which IMHO isn't that nice to trigger for every test.

Any chance we can push it to some extra wheel house e.g https://celery.github.io/celery-wheelhouse/repo/simple/ , since grpcio doesn't have binary releases for pypy ?

**Comment by thedrow** [comment]
@haimjether Please contribute the wheel in https://github.com/celery/celery-wheelhouse

**Review by Nusnus — COMMENTED** [review]
Any update on the PR? 🙏

**Comment by haimjether** [comment]
Added https://github.com/celery/celery-wheelhouse/pull/4 for the grpcio wheel - once it's merged the pypy tests should pass

**Comment by thedrow** [comment]
@haimjether The wheel is merged. Please rebase.

**Review by thedrow — CHANGES_REQUESTED** [review]
[Unit tests](https://github.com/celery/celery/actions/runs/11159409087/job/31017739244?pr=9302#step:7:993) are failing for 3.13.

**Comment by haimjether** [comment]
> [Unit tests](https://github.com/celery/celery/actions/runs/11159409087/job/31017739244?pr=9302#step:7:993) are failing for 3.13.
The issue is in lack of support of Google's Firestore for Python 3.13. 

Created an issue https://github.com/googleapis/python-firestore/issues/973

**Review by Nusnus — COMMENTED** [review]
If you want, you may disable 3.13 (remove from list in workflow) to see if the rest of the CI passes.
We can revert the commit before merge.

**Comment by haimjether** [comment]
@Nusnus ,  IMHO looks like everything but the rc python3.13 passes

**Comment by Nusnus** [comment]
> @Nusnus , IMHO looks like everything but the rc python3.13 passes

Isn’t it a real bug/error?
```
2024-10-13T17:45:32.6369811Z ==================================== ERRORS ====================================
2024-10-13T17:45:32.6370908Z _________________ ERROR collecting t/unit/backends/test_gcs.py _________________
2024-10-13T17:45:32.6371932Z t/unit/backends/test_gcs.py:7: in <module>
2024-10-13T17:45:32.6372576Z     from celery.backends.gcs import GCSBackend
2024-10-13T17:45:32.6374064Z celery/backends/gcs.py:30: in <module>
2024-10-13T17:45:32.6374991Z     from google.cloud import firestore, firestore_admin_v1
2024-10-13T17:45:32.6376422Z .tox/3.13-unit/lib/python3.13/site-packages/google/cloud/firestore_admin_v1/__init__.py:18: in <module>
2024-10-13T17:45:32.6377802Z     from .services.firestore_admin import FirestoreAdminClient
2024-10-13T17:45:32.6379129Z .tox/3.13-unit/lib/python3.13/site-packages/google/cloud/firestore_admin_v1/services/firestore_admin/__init__.py:16: in <module>
2024-10-13T17:45:32.6380386Z     from .client import FirestoreAdminClient
2024-10-13T17:45:32.6381639Z .tox/3.13-unit/lib/python3.13/site-packages/google/cloud/firestore_admin_v1/services/firestore_admin/client.py:53: in <module>
2024-10-13T17:45:32.6383052Z     from google.cloud.firestore_admin_v1.services.firestore_admin import pagers
2024-10-13T17:45:32.6384543Z .tox/3.13-unit/lib/python3.13/site-packages/google/cloud/firestore_admin_v1/services/firestore_admin/pagers.py:40: in <module>
2024-10-13T17:45:32.6385801Z     from google.cloud.firestore_admin_v1.types import field
2024-10-13T17:45:32.6386890Z .tox/3.13-unit/lib/python3.13/site-packages/google/cloud/firestore_admin_v1/types/__init__.py:19: in <module>
2024-10-13T17:45:32.6387911Z     from .database import (
2024-10-13T17:45:32.6388864Z .tox/3.13-unit/lib/python3.13/site-packages/google/cloud/firestore_admin_v1/types/database.py:34: in <module>
2024-10-13T17:45:32.6389745Z     class Database(proto.Message):
2024-10-13T17:45:32.6390729Z .tox/3.13-unit/lib/python3.13/site-packages/proto/message.py:279: in __new__
2024-10-13T17:45:32.6391696Z     file_info.generate_file_pb(new_class=cls, fallback_salt=full_name)
2024-10-13T17:45:32.6393277Z .tox/3.13-unit/lib/python3.13/site-packages/proto/_file_info.py:104: in generate_file_pb
2024-10-13T17:45:32.6394085Z     pool.Add(self.descriptor)
2024-10-13T17:45:32.6395176Z E   TypeError: Couldn't build proto file into descriptor pool: duplicate symbol 'google.firestore.admin.v1.Database.__firstlineno__’
```

**Comment by haimjether** [comment]
> @haimjether
> 
> > @Nusnus , IMHO looks like everything but the rc python3.13 passes
> 
> Isn’t it a real bug/error?
> 
> ```
> 2024-10-13T17:45:32.6369811Z ==================================== ERRORS ====================================
> 2024-10-13T17:45:32.6370908Z _________________ ERROR collecting t/unit/backends/test_gcs.py _________________
> 2024-10-13T17:45:32.6371932Z t/unit/backends/test_gcs.py:7: in <module>
> 2024-10-13T17:45:32.6372576Z     from celery.backends.gcs import GCSBackend
> 2024-10-13T17:45:32.6374064Z celery/backends/gcs.py:30: in <module>
> 2024-10-13T17:45:32.6374991Z     from google.cloud import firestore, firestore_admin_v1
> 2024-10-13T17:45:32.6376422Z .tox/3.13-unit/lib/python3.13/site-packages/google/cloud/firestore_admin_v1/__init__.py:18: in <module>
> 2024-10-13T17:45:32.6377802Z     from .services.firestore_admin import FirestoreAdminClient
> 2024-10-13T17:45:32.6379129Z .tox/3.13-unit/lib/python3.13/site-packages/google/cloud/firestore_admin_v1/services/firestore_admin/__init__.py:16: in <module>
> 2024-10-13T17:45:32.6380386Z     from .client import FirestoreAdminClient
> 2024-10-13T17:45:32.6381639Z .tox/3.13-unit/lib/python3.13/site-packages/google/cloud/firestore_admin_v1/services/firestore_admin/client.py:53: in <module>
> 2024-10-13T17:45:32.6383052Z     from google.cloud.firestore_admin_v1.services.firestore_admin import pagers
> 2024-10-13T17:45:32.6384543Z .tox/3.13-unit/lib/python3.13/site-packages/google/cloud/firestore_admin_v1/services/firestore_admin/pagers.py:40: in <module>
> 2024-10-13T17:45:32.6385801Z     from google.cloud.firestore_admin_v1.types import field
> 2024-10-13T17:45:32.6386890Z .tox/3.13-unit/lib/python3.13/site-packages/google/cloud/firestore_admin_v1/types/__init__.py:19: in <module>
> 2024-10-13T17:45:32.6387911Z     from .database import (
> 2024-10-13T17:45:32.6388864Z .tox/3.13-unit/lib/python3.13/site-packages/google/cloud/firestore_admin_v1/types/database.py:34: in <module>
> 2024-10-13T17:45:32.6389745Z     class Database(proto.Message):
> 2024-10-13T17:45:32.6390729Z .tox/3.13-unit/lib/python3.13/site-packages/proto/message.py:279: in __new__
> 2024-10-13T17:45:32.6391696Z     file_info.generate_file_pb(new_class=cls, fallback_salt=full_name)
> 2024-10-13T17:45:32.6393277Z .tox/3.13-unit/lib/python3.13/site-packages/proto/_file_info.py:104: in generate_file_pb
> 2024-10-13T17:45:32.6394085Z     pool.Add(self.descriptor)
> 2024-10-13T17:45:32.6395176Z E   TypeError: Couldn't build proto file into descriptor pool: duplicate symbol 'google.firestore.admin.v1.Database.__firstlineno__’
> ```

It is .
Google's firestore package is broken in Python 3.13 
I re-pinged google folks once again today  :shrug:

**Review by auvipy — COMMENTED** [review]
may be we could ignore/skip python 3.13 test for this PR as not all the dependencies of celery project still not working with python 3.13 yet?

**Comment by Nusnus** [comment]
> may be we could ignore/skip python 3.13 test for this PR as not all the dependencies of celery project still not working with python 3.13 yet?

I agree with @auvipy , let's do it

**Comment by haimjether** [comment]
Do I need to make this change as  part of the patch ?

**Comment by Nusnus** [comment]
@haimjether 
> Do I need to make this change as part of the patch ?

Hmm, you’re thinking about using a different PR so we can easily revert it later?
I like your idea haha

Sure go for it and ping me for a fast review.

**Comment by Nusnus** [comment]
@haimjether 
> Actually I meant should my PR change to to somehow ignore Python 3.13 😊

Yeah sure np just commit a “skip if py13” and link the relevant issue in the reason string.
Push to this PR & rebase/merge `main`.

Ping me if you need any help.

**Review by Nusnus — COMMENTED** [review]
The CI passes, thanks. I'll review tomorrow!

**Review by Nusnus — APPROVED** [review]
Thank you, looks great!

P.S
Do you think you can also cover the few remaining code coverage warnings?
Not critical though.

**Comment by Nusnus** [comment]
> LGTM

@auvipy please review the changes and approve/merge if it LGTyou2 haha

**Review by auvipy — APPROVED** [review]

**Review by benglewis — COMMENTED** [review]
I'm having some trouble with this PR's changes. I'd appreciate if it is possible to change it a little

File: docs/userguide/configuration.rst
Comment (haimjether) [reply docs/userguide/configuration.rst:1832]: @benglewis , it is required for the native chord support: https://github.com/celery/celery/pull/9302

Comment (benglewis) [reply docs/userguide/configuration.rst:1832]: I don't mean to be too stupid, but what exactly does that mean? 🙈 What are chords in the context of Celery? Why is this new? Were chords non-native before? Is it really necessary for them to be required now?

Comment (haimjether) [reply docs/userguide/configuration.rst:1832]: Chords have been an integral part of Celery since v2.3 :)


Please have a look here: 
https://docs.celeryq.dev/en/latest/userguide/canvas.html#id11

Native means that the backend implements the reference count in a very efficient way.

File: t/unit/backends/test_gcs.py
Comment (haimjether) [reply t/unit/backends/test_gcs.py:20]: You can upgrade to Python 3.13, without the GCS. Does that help ?

Or did you mean that you are  planning to use Celery with GCS and Pub/Sub with Python 3.13 ?

Comment (auvipy) [reply t/unit/backends/test_gcs.py:20]: I have opened a PR https://github.com/celery/celery/pull/9677 to fix this issue of testing python 3.13

**Pull request merged**
